torch/csrc/autograd/python_cpp_function.cpp

```diff
@@ -78,7 +78,10 @@ int THPCppFunction_traverse(PyObject* self, visitproc visit, void *arg)
 
 int THPCppFunction_clear(PyObject* self)
 {
-  ((THPCppFunction*)self)->cdata.reset();
+  auto f = (THPCppFunction*)self;
+  // Remove the weak ref of the c++ object if it exist
+  f->cdata->pyobj = nullptr;
```

Comment:
You need to do the same thing (clear the `pyobj`) in `THPCppFunction_dealloc` (or have `_dealloc` call `_clear`). `tp_clear` is only called when there GC is trying to break circular references. In the common case, when there are no circular references, only `tp_dealloc` is called